src/pages/images/actions/forms/UploadImageForm.tsx

```diff
@@ -0,0 +1,181 @@
+import { ChangeEvent, FC, useState } from "react";
+import { useEventQueue } from "context/eventQueue";
+import { useFormik } from "formik";
+import { useToastNotification } from "context/toastNotificationProvider";
+import { createImageAlias, createImageFromTarball } from "api/images";
+import {
+  ActionButton,
+  Button,
+  Form,
+  Input,
+  Modal,
+} from "@canonical/react-components";
+import * as Yup from "yup";
+import { Link } from "react-router-dom";
+import { humanFileSize } from "util/helpers";
+import ProgressBar from "components/ProgressBar";
+import { UploadState } from "types/storage";
+import Loader from "components/Loader";
+import { useQueryClient } from "@tanstack/react-query";
+import { queryKeys } from "util/queryKeys";
+
+interface Props {
+  close: () => void;
+  project: string;
+}
+
+const UploadImageForm: FC<Props> = ({ close, project }) => {
+  const eventQueue = useEventQueue();
+  const toastNotify = useToastNotification();
+  const [uploadState, setUploadState] = useState<UploadState | null>(null);
+  const queryClient = useQueryClient();
+
+  const notifySuccess = () => {
+    const uploaded = <Link to={`/ui/project/${project}/images`}>uploaded</Link>;
+    toastNotify.success(<>Image {uploaded}.</>);
+  };
+
+  const changeFile = (e: ChangeEvent<HTMLInputElement>) => {
+    if (e.target.files) {
+      void formik.setFieldValue("fileList", e.target.files);
+    }
+  };
+
+  const getImageToTarballBody = (fileList: FileList): File | FormData => {
+    if (fileList.length != 1) {
+      // Sorting by Size. The metadata file is very likely to be smaller than the image itself.
+      const formData = new FormData();
+      const sortedFiles = Array.from(fileList).sort((a, b) => a.size - b.size);
+
+      formData.append("metadata", sortedFiles[0]);
+      formData.append("rootfs.img", sortedFiles[1]);
+
+      if (!formData.has("metadata") || !formData.has("rootfs.img")) {
+        throw new Error(
+          "Both metadata (tar.xz / gzip) and rootfs.img files are required",
+        );
+      }
+
+      return formData;
+    }
+    return fileList[0];
+  };
+
+  const formik = useFormik<{
+    alias?: string;
```

Comment:
```suggestion
    alias: string;
```
I think this is always string as it initializes with empty string.